packages/runed/src/lib/utilities/NetworkStatus/NetworkStatus.svelte.ts

```diff
@@ -0,0 +1,159 @@
+import { onMount } from "svelte";
+import { browser } from "$lib/internal/utils/browser.js";
+import { addEventListener } from "$lib/internal/utils/event.js";
+import { IsSupported } from "$lib/utilities/index.js";
+
+/**
+ * @desc The `NetworkInformation` interface of the Network Information API
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/NetworkInformation
+ */
+type NetworkInformation = {
+	readonly downlink: number;
+	readonly downlinkMax: number;
+	readonly effectiveType: "slow-2g" | "2g" | "3g" | "4g";
+	readonly rtt: number;
+	readonly saveData: boolean;
+	readonly type:
+		| "bluetooth"
+		| "cellular"
+		| "ethernet"
+		| "none"
+		| "wifi"
+		| "wimax"
+		| "other"
+		| "unknown";
+} & EventTarget;
+
+type NavigatorWithConnection = Navigator & { connection: NetworkInformation };
+
+/**
+ * Tracks the state of browser's network connection.
+ *
+ * @see {@link https://runed.dev/docs/utilities/network-status}
+ */
+export class NetworkStatus {
+	#isSupported = new IsSupported(() => browser && "navigator" in window);
+	#navigator?: Navigator = $derived(this.#isSupported.current ? window.navigator : undefined);
+	#connection?: NetworkInformation = $derived(
+		this.#navigator && "connection" in this.#navigator
+			? (this.#navigator as NavigatorWithConnection).connection
+			: undefined
+	);
+	#online: boolean = $state(false);
+	#updatedAt: Date = $state(new Date());
+	#downlink?: NetworkInformation["downlink"] = $state();
```

Comment:
Can't these be deriveds that depend on connection?

Reply:
Actually, can't we just use gets that use connection? Why do we need these individual fields?

Reply:
Can we get the updated state without these? If `connection` or its properties, or `online`/`offline` status changes (e.g., random internet disconnection), how will the client be notified?

Reply:
connection can be a state, since all of the other ones are just set from it. then `updateStatus` updates connection, and that's it

Reply:
I pushed the commit where I simplified the state by depending solely on the derived connection. Seems like it's working exactly the same as before so definitely a great suggestion :)